src/opencl/axcrypt_kernel.cl

```diff
@@ -0,0 +1,115 @@
+/*
+ * This software is Copyright (c) 2018 Dhiru Kholia, Copyright (c) 2016
+ * Fist0urs <eddy.maaalou at gmail.com>, Copyright (c) 2018 magnum, and it is
+ * hereby released to the general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ */
+
+#include "opencl_device_info.h"
+#include "opencl_misc.h"
+#include "opencl_sha1_ctx.h"
+#include "opencl_aes.h"
+
+#ifndef PLAINTEXT_LENGTH
+#error PLAINTEXT_LENGTH must be defined
+#endif
+
+// input
+typedef struct {
+	uint length;
+	uchar v[PLAINTEXT_LENGTH];
+} axcrypt_password;
+
+// input
+typedef struct {
+	uint version;
+        uint key_wrapping_rounds;
+        uint keyfile_length;
+        uchar salt[16];
+        uchar wrappedkey[24];
+        uchar keyfile[4096];
+} axcrypt_salt;
+
+// output
+typedef struct {
+	uint cracked;
+} axcrypt_out;
+
+#define PUT_64BITS_XOR_MSB(cp, value) ( \
+                (cp)[0] ^= (unsigned char)((value)), \
+                (cp)[1] ^= (unsigned char)((value) >> 8), \
+                (cp)[2] ^= (unsigned char)((value) >> 16), \
+                (cp)[3] ^= (unsigned char)((value) >> 24 ) )
+
+inline int axcrypt_decrypt(__global const axcrypt_password *inbuffer, uint gid, __constant axcrypt_salt *cur_salt, __global axcrypt_out *output)
+{
+	uchar password[PLAINTEXT_LENGTH];
+	uchar keyfile[4096];
+	int password_length = inbuffer[gid].length;
+	uchar salt[16];
+	unsigned char KEK[20], lsb[24], cipher[16];
+	AES_KEY akey;
+	SHA_CTX ctx;
+	uint i;
+	int j, nb_iterations = cur_salt->key_wrapping_rounds;
+
+	memcpy_gp(password, inbuffer[gid].v, password_length);
+	memcpy_macro(salt, cur_salt->salt, 16);
+	if (cur_salt->keyfile_length != 0)
+		memcpy_macro(keyfile, cur_salt->keyfile, cur_salt->keyfile_length);
+
+	SHA1_Init(&ctx);
+	SHA1_Update(&ctx, password, password_length);
+	if (cur_salt->keyfile_length != 0) {
+		SHA1_Update(&ctx, keyfile, cur_salt->keyfile_length);
+	}
+	SHA1_Final(KEK, &ctx);
+
+	/* hash XOR salt => KEK */
+	for (i = 0; i < sizeof(cur_salt->salt); i++)
+		KEK[i] ^= salt[i];
+
+	memcpy_macro(lsb, cur_salt->wrappedkey + 8, 16);
+
+	AES_set_decrypt_key(KEK, 128, &akey);
+
+	/* set msb */
+	memcpy_macro(cipher, cur_salt->wrappedkey, 8);
+
+	/* custom AES un-wrapping loop */
+	for (j = nb_iterations - 1; j >= 0; j--) {
+		/* 1st block treatment */
+		/* MSB XOR (NUMBER_AES_BLOCKS * j + i) */
+		PUT_64BITS_XOR_MSB(cipher, 2 * j + 2);
+		/* R[i] */
+		memcpy_macro(cipher + 8, lsb + 8, 8);
+		/* AES_ECB(KEK, (MSB XOR (NUMBER_AES_BLOCKS * j + i)) | R[i]) */
+		AES_decrypt(cipher, cipher, &akey);
+		memcpy_macro(lsb + 8, cipher + 8, 8);
+
+		/* 2nd block treatment */
+		PUT_64BITS_XOR_MSB(cipher, 2 * j + 1);
+		memcpy_macro(cipher + 8, lsb, 8);
+		AES_decrypt(cipher, cipher, &akey);
+		memcpy_macro(lsb, cipher + 8, 8);
+	}
```

Comment:
This hot loop should probably be optimized - make `cipher` and `lsb` unions that also have 32-bit members, then replace all `memcpy_macro()` with 32-bit copies eg.
```diff
- 		memcpy_macro(lsb, cipher + 8, 8);
+ 		lsb.w[0] = cipher.w[2];
+ 		lsb.w[1] = cipher.w[3];
```
You could also write a 32-bit version of `PUT_64BITS_XOR_MSB()`

Reply:
Doing so might not be straightforward -> `cipher` and `lsb` need to hold different values before the loop starts.

Reply:
Those operations can be left as-is (except for changing eg. `lsb` to `lsb.b`